C3: Automatically use the latest version of c3 if not running latest

jculvey opened the pull request:
Fixes #3793.

**What this PR solves / how to test:**

Automatically updates to the latest version of c3 when a newer version is available.

**Associated docs issue(s)/PR(s):**

- [insert associated docs issue(s)/PR(s)]

**Author has included the following, where applicable:**

- [ ] Tests
- [ x ] Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))

**Reviewer is to perform the following, as applicable:**

- Checked for inclusion of relevant tests
- Checked for inclusion of a relevant changeset
- Checked for creation of associated docs updates
- Manually pulled down the changes and spot-tested

**Note for PR author:**

We want to celebrate and highlight awesome PR review! If you think this PR received a particularly high-caliber review, please assign it the label `highlight pr review` so future reviewers can take inspiration and learn from it.

dario-piotrowicz commented on packages/create-cloudflare/src/cli.ts:
```diff
@@ -30,6 +32,39 @@ const WRANGLER_DEFAULTS = {
+
+	// Use a spinner when running this check since it may take some time
+	const s = spinner(spinnerFrames.vertical);
+	s.start("Checking for newer version");
```
```suggestion
	s.start("Checking for updates");
```
? 🤔

jculvey replied:
Updates feels weird, how about "Checking if a newer version is available"?

dario-piotrowicz replied:
Nah don't worry checking for newer version is fine... I just feel that it sounded a bit specific, but I'm totally fine with it if you prefer it 🙂👍

dario-piotrowicz commented on packages/create-cloudflare/src/cli.ts:
```diff
@@ -30,6 +32,39 @@ const WRANGLER_DEFAULTS = {
+
+export const runLatest = async () => {
+	const args = process.argv.slice(2);
+	await runCommand(`npm create cloudflare@latest ${args.join(" ")}`);
```
shouldn't this use the current package manager instead of hardcoding `npm`?

IgorMinar replied:
I think so too.

jculvey replied:
Good catch!

IgorMinar commented on packages/create-cloudflare/src/cli.ts:
```diff
@@ -30,6 +32,39 @@ const WRANGLER_DEFAULTS = {
+	}
+
+	// Use a spinner when running this check since it may take some time
+	const s = spinner(spinnerFrames.vertical);
```
why different style spinner for this one? is there a semantic difference between waiting on this step vs waiting on other steps that you are trying to convey?

jculvey replied:
The current spinner uses the same characters that we use for starting, stopping, and continuing sections (the top and bottom corners, and the gray bars to the left). It feels a little out of place when not inside one of those contexts

IgorMinar replied:
I see. ok. sounds good.

IgorMinar commented on packages/create-cloudflare/src/cli.ts:
```diff
@@ -30,6 +32,39 @@ const WRANGLER_DEFAULTS = {
+};
+
+const isUpdateAvailable = async () => {
+	if (process.env.VITEST) {
```
are you hardcoding this to make our own e2e test suite work? could we do something more generic like `CI` or use both VITEST and CI?

CI use used by almost everyone: https://github.com/orgs/community/discussions/26628

jculvey replied:
I've had the same thought but I've been holding off until I had a good first use case for "CI" -- meaning something I only want to happen in github actions but not on dev laptops.

For example we'd want skip auto-updates when running the tests locally, and by checking VITEST which gets set automatically by the runner, devs can't make a mistake of forgetting to set CI in a test run and seeing inconsistent results.

IgorMinar commented on packages/create-cloudflare/src/cli.ts:
```diff
@@ -30,6 +32,39 @@ const WRANGLER_DEFAULTS = {
+};
+
+const isUpdateAvailable = async () => {
+	if (process.env.VITEST) {
```
```suggestion
	if (process.env.CI || process.env.VITEST) {
```

IgorMinar replied:
And please add some docs explaining what this is about. thank you

IgorMinar commented on packages/create-cloudflare/src/cli.ts:
```diff
@@ -91,6 +126,7 @@ export const parseArgs = async (argv: string[]): Promise<Partial<C3Args>> => {
 		.option("type", { type: "string" })
 		.option("framework", { type: "string" })
 		.option("deploy", { type: "boolean" })
+		.option("auto-update", { type: "boolean", default: C3_DEFAULTS.autoUpdate })
```
can you please add description for this field too, otherwise `--help` output won't be very useful. 

could you also backfill `descriptions` for the other flags that don't have one? thanks

IgorMinar commented:
> LGTM! Thank you, James!

oh uh, I see the e2e tests failing, please take a look

jculvey commented:
> Are there going to be any breaking changes when the next major version of `c3` is published? It's probably too early to say, but there likely will be and doing this will almost certainly break someone's workflow in future.
> Perhaps this could ensure more stability if it only used the latest version of the current major version, instead of the latest version in general?

What we're seeing today is that many users get stuck with cached, older versions of `create-cloudflare` as the most common workflow is running `npm create cloudflare` to scaffold a new application, and perhaps running it days/weeks/months in the future for the same use case. I'm sure too at some point we will create a truly breaking change and bump to a major version, but I think for this use case we still want users to have the latest version.

If you're using c3 as part of an automated workflow, you can always use the `--no-auto-update` flag to skip this check entirely. With that, you should be able to explicitly control the version. I've also added a check that will skip auto-updating when c3 is executing in a non-interactive context.

 Do you have a particular workflow in mind that you're worried about?

Cherry commented:
> I've also added a check that will skip auto-updating when c3 is executing in a non-interactive context.

This makes a lot of sense!

> What we're seeing today is that many users get stuck with cached, older versions of `create-cloudflare` as the most common workflow is running `npm create cloudflare` to scaffold a new application, and perhaps running it days/weeks/months in the future for the same use case.

I totally understand this - `npx` / `npm` cache can be a very weird thing, and outside of doing `@latest`, it often uses old versions. 

> Do you have a particular workflow in mind that you're worried about?

My concern is simply the major version bump happening without my knowledge, and changing things that I might not expect. For example, let's say today I create a new project using `c3` and it gives me a particular file structure, versions, etc. but then tomorrow a new major release comes out that changes the file/folder structure, I run it again for a new project, and now the output is entirely different from what I saw yesterday. That's a pretty disjunct experience as a developer and not one I'm likely to even realise happened immediately.

And another example would be testing/validation. Let's say I come across a bug in a recent release of `c3` and want to test on a previous version to validate if it's a new bug, a regression, etc. I could previously run `npm create cloudflare@2.1.0` (etc.) and validate this quickly, whereas now I would have to do `--no-auto-update`. That in and of itself feels like a breaking change to me which might deserve a `3.x` here. 🤔

Resolving the issue with older versions being cached and used as preference might outweigh those concerns, and that's fine, but I wanted to raise them as someone who has dealt with software automatically upgrading major versions before and being bitten by it so many times. Auto-upgrading on patch and even minor releases is usually fine, but auto-upgrading on major versions is just asking for trouble, from my experience.

isaac-mcfadyen commented:
Perhaps a warning could also be considered?

I'd agree with not wanting to auto-update things. I don't like when my tools auto-update without me specifying they should do so, because then the flags/options might change without me knowing.

Rather, I'd prefer a warning. Something like:
`WARN: you are not using the latest version of create-cloudflare. Upgrade to <x> for the latest features and bug fixes`

That way I can go:
> Oh, I'm not running latest, things might break when I upgrade but I have lots of time to troubleshoot/look at the changes right now

*OR*

> I don't have time to upgrade right now, I just need to get this project done ASAP, I'll upgrade later when I have time to look at the changes

jculvey commented:
@Cherry @isaac-mcfadyen thanks for the feedback. I've updated the PR to skip auto-updating to new major versions. I've also adjusted the changeset so to be a major version bump.

Cherry commented:
Thanks @jculvey ❤️

IgorMinar commented:
I'm just catching up with this PR! thanks @Cherry for the feedback and @jculvey for the changes.

I do however have a concern about making this a major version change. I can see how one could argue the this behavior is a breaking change, but so could be said about so many things in c3 since it's a meta-meta-cli and the underlying CLIs can easily introduce a breaking change which we simply pass through.

I would also argue that some bugs we've fixed recently could in some rare circumstances represent a breaking change, for example: https://github.com/cloudflare/workers-sdk/pull/3807

Here is however the main problem, I see an increasing number of documentation and blog posts that instruct users to use `npm create cloudflare@2` with `@2` being the significant bit. Quick search highlights these docs:
- https://blog.logrocket.com/integrating-supabase-cloudflare-workers/
- https://zenn.dev/kentarok/scraps/032a9c3e256522
- https://dev.to/seratch/running-slack-app-on-cloudflare-workers-3hhn
- https://stackoverflow.com/questions/76027117/how-to-clone-an-existing-cloudflare-worker
- and many others

So if we bump the major version due to this single change, all of these docs will suddenly be obsolete and encourage users to install the wrong version of the package.

This is bad for the users and also for us because we'll fragment the user-base between v2 and v3 users.

This is also unnecessary. The risks associated with releasing this a minor version are minimal and easily justifiable now that we made the autoupdate to work only over minor versions and is disabled by default in non-interactive mode.

Cherry commented:
I would agree. While _technically_ a breaking change, the risks here are now minimal with the restriction preventing major versions from being auto-upgraded to, as long as any future changes that would have more major impact are treated as breaking changes.

I'd +1 a semver-minor change here instead of semver-major with the new upgrade restrictions in place.
